<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Thirty datasets with inapplicable data — inapplicable.datasets • TreeSearch</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous"><link href="../docsearch.css" rel="stylesheet"><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Thirty datasets with inapplicable data — inapplicable.datasets"><meta property="og:description" content="These are the datasets used to evaluate the behaviour of the inapplicable
algorithm in Brazeau et al. (2019)
.
The name of each item corresponds to the datasets listed below.
Datasets are sorted into two subsets, each sorted alphabetically;
the first subset comprise simpler datasets with faster processing times.
inapplicable.datasets provide the data in the matrix format generated by
read.nexus.data(); inapplicable.phyData are in phyDat format.
inapplicable.trees lists for each dataset a sample of up to 50 trees
obtained by tree search under each inapplicable treatment, named accordingly.
inapplicable.citations is a named character vector specifying the source of
each dataset."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=312413071"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '312413071');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TreeSearch</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/custom.html">Tree search with custom optimality criteria</a>
    </li>
    <li>
      <a href="../articles/getting-started.html">Getting started: Installing R and TreeSearch</a>
    </li>
    <li>
      <a href="../articles/profile-scores.html">Profile parsimony</a>
    </li>
    <li>
      <a href="../articles/profile.html">Tree Search with Profile Parsimony</a>
    </li>
    <li>
      <a href="../articles/tree-search.html">Getting started: Simple tree searches</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ms609/TreeSearch/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul><form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off"></div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Thirty datasets with inapplicable data</h1>
    <small class="dont-index">Source: <a href="https://github.com/ms609/TreeSearch/blob/HEAD/R/data.R" class="external-link"><code>R/data.R</code></a></small>
    <div class="hidden name"><code>inapplicable.datasets.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>These are the datasets used to evaluate the behaviour of the inapplicable
algorithm in Brazeau et al. (2019)
.
The name of each item corresponds to the datasets listed below.
Datasets are sorted into two subsets, each sorted alphabetically;
the first subset comprise simpler datasets with faster processing times.
<code>inapplicable.datasets</code> provide the data in the matrix format generated by
<code><a href="https://rdrr.io/pkg/ape/man/read.nexus.data.html" class="external-link">read.nexus.data()</a></code>; <code>inapplicable.phyData</code> are in <code>phyDat</code> format.
<code>inapplicable.trees</code> lists for each dataset a sample of up to 50 trees
obtained by tree search under each inapplicable treatment, named accordingly.
<code>inapplicable.citations</code> is a named character vector specifying the source of
each dataset.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="va">inapplicable.datasets</span>

<span class="va">inapplicable.phyData</span>

<span class="va">inapplicable.trees</span>

<span class="va">inapplicable.citations</span></code></pre></div>
    </div>

    <div id="format">
    <h2>Format</h2>
    <p>An object of class <code>list</code> of length 30.</p>
<p>An object of class <code>list</code> of length 30.</p>
<p>An object of class <code>list</code> of length 31.</p>
<p>An object of class <code>character</code> of length 30.</p>
    </div>
    <div id="source">
    <h2>Source</h2>
    <p>Subset one (faster processing):</p><dl><dt>Agnarsson2004</dt>
<dd><p>AGNARSSON, I. 2004. Morphological phylogeny of cobweb spiders and their
relatives (Araneae, Araneoidea, Theridiidae). Zoological Journal of the
Linnean Society, 141, 447--626.</p></dd><dt>Capa2011</dt>
<dd><p>CAPA, M., HUTCHINGS, P., AGUADO, M. T. and BOTT, N. J. 2011. Phylogeny of
Sabellidae (Annelida) and relationships with other taxa inferred from
morphology and multiple genes. Cladistics, 27, 449--469.</p></dd><dt>DeAssis2011</dt>
<dd><p>DE ASSIS, J. E. and CHRISTOFFERSEN, M. L. 2011. Phylogenetic relationships
within Maldanidae (Capitellida, Annelida), based on morphological characters.
Systematics and Biodiversity, 9, 233--245.</p></dd><dt>OLeary1999</dt>
<dd><p>O'LEARY, M. A. and GEISLER, J. H. 1999. The position of Cetacea within
Mammalia: phylogenetic analysis of morphological data from extinct and
extant taxa. Systematic Biology, 48, 455--490.</p></dd><dt>Rousset2004</dt>
<dd><p>ROUSSET, V., ROUSE, G. W., SIDDALL, M. E., TILLIER, A. and PLEIJEL, F.
2004. The phylogenetic position of Siboglinidae (Annelida) inferred from
18S rRNA, 28S rRNA and morphological data. Cladistics, 20, 518--533.</p></dd><dt>Sano2011</dt>
<dd><p>SANO, M. and AKIMOTO, S.-I. 2011. Morphological phylogeny of
gall-forming aphids of the tribe Eriosomatini (Aphididae: Eriosomatinae).
Systematic Entomology, 36, 607--627.</p></dd><dt>Sansom2010</dt>
<dd><p>SANSOM, R. S., FREEDMAN, K., GABBOTT, S. E., ALDRIDGE, R. J. and
PURNELL, M. A. 2010. Taphonomy and affinity of an enigmatic Silurian vertebrate,
<em>Jamoytius kerwoodi</em> White. Palaeontology, 53, 1393--1409.</p></dd><dt>Schulze2007</dt>
<dd><p>SCHULZE, A., CUTLER, E. B. and GIRIBET, G. 2007. Phylogeny of sipunculan
worms: A combined analysis of four gene regions and morphology. Molecular Phylogenetics and Evolution, 42, 171--92.</p></dd><dt>Shultz2007</dt>
<dd><p>SHULTZ, J. W. 2007. A phylogenetic analysis of the arachnid orders based
on morphological characters. Zoological Journal of the Linnean Society, 150, 221--265.</p></dd><dt>Wetterer2000</dt>
<dd><p>WETTERER, A. L., ROCKKMAN, M. V. and SIMMONS, N. B. 2000. Phylogeny of
phyllostomid bats (Mammalia: Chiroptera): data from diverse morphological
systems, sex chromosomes, and restriction sites. Bulletin of the American Museum of Natural History, 248, 1--200.</p></dd><dt>Wills2012</dt>
<dd><p>WILLS, M. A., GERBER, S., RUTA, M. and HUGHES, M. 2012. The disparity of
priapulid, archaeopriapulid and palaeoscolecid worms in the light of new data.
Journal of Evolutionary Biology, 25, 2056--2076.</p></dd></dl>Subset two (longer processing times):<dl><dt>Aguado2009</dt>
<dd><p>AGUADO, M. T. and SAN MARTIN, G. 2009. Phylogeny of Syllidae (Polychaeta)
based on morphological data. Zoologica Scripta, 38, 379--402.</p></dd><dt>Aria2015</dt>
<dd><p>ARIA, C., CARON, J. B. and GAINES, R. 2015. A large new leanchoiliid from
the Burgess Shale and the influence of inapplicable states on stem arthropod
phylogeny. Palaeontology, 58, 629--660.</p></dd><dt>Asher2005</dt>
<dd><p>ASHER, R. J. and HOFREITER, M. 2006. Tenrec phylogeny and the noninvasive
extraction of nuclear DNA. Systematic biology, 55, 181--94.</p></dd><dt>Baker2009</dt>
<dd><p>BAKER, W. J., SAVOLAINEN, V., ASMUSSEN-LANGE, C. B., CHASE, M. W.,
DRANSFIELD, J., FOREST, F., HARLEY, M. M., UHL, N. W. and WILKINSON, M. 2009.
Complete generic-level phylogenetic analyses of palms (Arecaceae) with
comparisons of supertree and supermatrix approaches.
Systematic Biology, 58, 240--256.</p></dd><dt>Bouchenak2010</dt>
<dd><p>BOUCHENAK-KHELLADI, Y., VERBOOM, G. A., SAVOLAINEN, V. and HODKINSON, T. R.
2010. Biogeography of the grasses (Poaceae): a phylogenetic approach to reveal
evolutionary history in geographical space and geological time. Botanical
Journal of the Linnean Society, 162, 543--557.</p></dd><dt>Conrad2008</dt>
<dd><p>CONRAD, J. L. 2008. Phylogeny And Systematics Of Squamata (Reptilia) Based
On Morphology. Bulletin of the American Museum of Natural History, 310, 1--182.</p></dd><dt>Dikow2009</dt>
<dd><p>DIKOW, T. 2009. A phylogenetic hypothesis for Asilidae based on a total
evidence analysis of morphological and DNA sequence data (Insecta: Diptera:
Brachycera: Asiloidea). Organisms Diversity and Evolution, 9, 165--188.</p></dd><dt>Eklund2004</dt>
<dd><p>EKLUND, H., DOYLE, J. A. and HERENDEEN, P. S. 2004. Morphological
phylogenetic analysis of living and fossil Chloranthaceae. International
Journal of Plant Sciences, 165, 107--151.</p></dd><dt>Geisler2001</dt>
<dd><p>GEISLER, J. H. 2001. New morphological evidence for the phylogeny of
Artiodactyla, Cetacea, and Mesonychidae. American Museum Novitates, 3344, 53.</p></dd><dt>Giles2015</dt>
<dd><p>GILES, S., FRIEDMAN, M. and BRAZEAU, M. D. 2015. Osteichthyan-like cranial
conditions in an Early Devonian stem gnathostome. Nature, 520, 82--85.</p></dd><dt>Griswold1999</dt>
<dd><p>GRISWOLD, C. E., CODDINGTON, J. A., PLATNICK, N. I. and FORSTER, R. R. 1999.
Towards a phylogeny of entelegyne spiders (Araneae, Araneomorphae, Entelegynae).
Journal of Arachnology, 27, 53--63.</p></dd><dt>Liljeblad2008</dt>
<dd><p>LILJEBLAD, J., RONQUIST, F., NIEVES-ALDREY, J. L., FONTAL-CAZALLA, F.,
ROS-FARRE, P., GAITROS, D. and PUJADE-VILLAR, J. 2008. A fully web-illustrated
morphological phylogenetic study of relationships among oak gall wasps and
their closest relatives (Hymenoptera: Cynipidae).</p></dd><dt>Loconte1991</dt>
<dd><p>LOCONTE, H. and STEVENSON, D. W. 1991. Cladistics of the Magnoliidae.
Cladistics, 7, 267--296.</p></dd><dt>Longrich2010</dt>
<dd><p>LONGRICH, N. R., SANKEY, J. and TANKE, D. 2010. <em>Texacephale langstoni</em>,
a new genus of pachycephalosaurid (Dinosauria: Ornithischia) from the upper
Campanian Aguja Formation, southern Texas, USA. Cretaceous Research, 31, 274--284.</p></dd><dt>OMeara2014</dt>
<dd><p>O'MEARA, R. N. and THOMPSON, R. S. 2014. Were There Miocene Meridiolestidans?
Assessing the phylogenetic placement of <em>Necrolestes patagonensis</em> and the
presence of a 40 million year Meridiolestidan ghost lineage. Journal
of Mammalian Evolution, 21, 271--284.</p></dd><dt>Rougier2012</dt>
<dd><p>ROUGIER, G. W., WIBLE, J. R., BECK, R. M. D. and APESTEGUIA, S. 2012.
The Miocene mammal <em>Necrolestes</em> demonstrates the survival of a
Mesozoic nontherian lineage into the late Cenozoic of South America.
Proceedings of the National Academy of Sciences, 109, 20053--8.</p></dd><dt>Sharkey2011</dt>
<dd><p>SHARKEY, M. J., CARPENTER, J. M., VILHELMSEN, L., HERATY, J., LILJEBLAD, J.,
DOWLING, A. P. G., SCHULMEISTER, S., MURRAY, D., DEANS, A. R., RONQUIST, F.,
KROGMANN, L. and WHEELER, W. C. 2012. Phylogenetic relationships among
superfamilies of Hymenoptera. Cladistics, 28, 80--112.</p></dd><dt>Sundue2010</dt>
<dd><p>SUNDUE, M. A., ISLAM, M. B. and RANKER, T. A. 2010. Systematics of
Grammitid Ferns (Polypodiaceae): Using Morphology and Plastid Sequence Data
to Resolve the Circumscriptions of Melpomene and the Polyphyletic Genera
<em>Lellingeria</em> and <em>Terpsichore</em>. Systematic Botany, 35, 701--715.</p></dd><dt>Vinther2008</dt>
<dd><p>VINTHER, J., VAN ROY, P. and BRIGGS, D. E. G. 2008. Machaeridians are
Palaeozoic armoured annelids. Nature, 451, 185--188.</p></dd><dt>Wilson2003</dt>
<dd><p>WILSON, G. D. F. and EDGECOMBE, G. D. 2003. The Triassic isopod
<em>Protamphisopus wianamattensis</em> (Chilton) and comparison by extant
taxa (Crustacea, Phreatoicidea). Journal of Paleontology, 77, 454--470.</p></dd><dt>Wortley2006</dt>
<dd><p>WORTLEY, A. H. and SCOTLAND, R. W. 2006. The effect of combining molecular
and morphological data in published phylogenetic analyses.
Systematic Biology, 55, 677--685.</p></dd><dt>Zanol2014</dt>
<dd><p>ZANOL, J., HALANYCH, K. M. and FAUCHALD, K. 2014. Reconciling taxonomy and
phylogeny in the bristleworm family Eunicidae (Polychaete, Annelida).
Zoologica Scripta, 43, 79--100.</p></dd><dt>Zhu2013</dt>
<dd><p>ZHU, M., YU, X., AHLBERG, P. E., CHOO, B., LU, J., QIAO, T., QU, Q., ZHAO, W.,
JIA, L., BLOM, H. and ZHU, Y. 2013. A Silurian placoderm with osteichthyan-like
marginal jaw bones. Nature, 502, 188--193.</p></dd></dl></div>
    <div id="references">
    <h2>References</h2>
    <p>Brazeau MD, Guillerme T, Smith MR (2019).
“An algorithm for morphological phylogenetic analysis with inapplicable data.”
<em>Systematic Biology</em>, <b>68</b>(4), 619--631.
<a href="https://doi.org/10.1093/sysbio/syy083" class="external-link">doi:10.1093/sysbio/syy083</a>
.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">'inapplicable.datasets'</span>, package <span class="op">=</span> <span class="st">'TreeSearch'</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">inapplicable.datasets</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "Agnarsson2004" "Aguado2009"    "Aria2015"      "Asher2005"    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5] "Capa2011"      "Conrad2008"    "DeAssis2011"   "Dikow2009"    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9] "Eklund2004"    "Geisler2001"   "Giles2015"     "Griswold1999" </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13] "Liljeblad2008" "Loconte1991"   "Longrich2010"  "OLeary1999"   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [17] "OMeara2014"    "Rougier2012"   "Rousset2004"   "Sano2011"     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [21] "Sansom2010"    "Schulze2007"   "Shultz2007"    "Vinther2008"  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [25] "Wetterer2000"  "Wills2012"     "Wilson2003"    "Wortley2006"  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [29] "Zanol2014"     "Zhu2013"      </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://smithlabdurham.github.io/" class="external-link">Martin R. Smith</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer></div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '8422cbc6478f9d3278ad355b48435342',
    indexName: 'treesearch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script></body></html>

